[SYCL] Implement hierarchical parallelism API.

kbobrovs opened the pull request:
This is the first part of SYCL hierarchical parallelism implementation. It
implements all related APIs:

- h_item class
- group::parallel_for_work_item functions
- handler::parallel_for_work_group functions

It is able to run workloads which use these APIs but do not contain data
or code with group-visible side effects between the work group and work
item scopes. Upcoming patches will implement the remaining part: private_data, data and code work-group level "scoping".

Signed-off-by: Konstantin Bobrovsky konstantin.s.bobrovsky@intel.com

agozillon commented on sycl/include/CL/sycl/detail/common.hpp:
```diff
@@ -113,6 +113,94 @@ const cl_mem_fence_flags CLK_GLOBAL_MEM_FENCE  = 0x02;
+// is to better support handling of situations where there must be a loop nest
+// over a multi-dimensional space - it allows to avoid generating unnecessary
+// outer loops like 'for (int z=0; z<1; z++)' in case of 1D and 2D iteration
+// spaces or writing specializations of the algorighms for 1D, 2D and 3D cases.
```
Not particularly relevant as it's understandable either way but "algorighms" -> "algorithms"

kbobrovs replied:
Thanks for the catch. Will change

agozillon commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -619,93 +592,109 @@ class handler {
-  template <typename KernelType, int Dims>
-  void parallel_for(nd_range<Dims> ExecutionRange, KernelType KernelFunc) {
-    parallel_for<KernelType, KernelType, Dims>(ExecutionRange, KernelFunc);
+  template <typename KernelName = std::nullptr_t, typename KernelType, int Dims>
```
Possibly a reason behind it that I'm not seeing at the moment, but is there a reason the dimension is named "Dims" here and "dimensions" in kernel_parallel_for_work_group and in the other parallel_for_work_group interface (https://github.com/intel/llvm/pull/221/files#diff-234ddc8d342a3b24bb61d130262e9624R665). Perhaps worth renaming the other two that use "dimensions" to "Dims" to keep consistency.

kbobrovs replied:
No particular reason. Will make it "Dims" - I agree, should be uniform.

agozillon commented on sycl/include/CL/sycl/detail/spirv_glue.hpp:
```diff
@@ -0,0 +1,60 @@
+
+#include "CL/__spirv/spirv_vars.hpp"
+
+#define DEFINE_INIT_SIZES(POSTFIX)                                             \
```
This may be a silly request, but would it be possible to call this something other than spirv_glue.hpp, as at the moment we currently use these functions as a sort of top level API that offloads to SPIRV/SPIR intrinsics based on what we compile for. e.g: https://github.com/triSYCL/sycl/blob/sycl/unified/next/sycl/include/CL/__spir/spir_vars.hpp I'm aware the names are SPIRV oriented though.

Maybe this isn't the place for that though or the best implementation on our end (I did have a slightly more complex/generic implementation in a prior commit but it was a bit overly complex for our current needs), but it would be nice to have a sort of top level id/size/group etc. API that calls the underlying implementation be it OpenCL SPIR calls, SPIRV calls or something else entirely. Hopefully this is a somewhat reasonable sounding idea/request.

kbobrovs replied:
Good point. We should not be tied to SPIRV, as there can be other models or even runtimes on the device side underneath. Is it correct interpretation of your concern? How about device_rt_intrin.hpp or device_rt_builtins.hpp? For now there is __spirv* variant only, later implementation for more native RTs can be added. @bader, please also comment.

agozillon replied:
That is indeed a correct interpretation of my concern! device_rt_builtins.hpp sounds fine by me but if you want to call it device_rt_intrin.hpp that's also all good in my book. Thanks for taking my concern into consideration!

bader replied:
I'm totally agree with @agozillon that `spirv_glue` is unfortunate name - even w/o applying his "generalization idea".
I think current implementation of these functions can reside in `spirv_vars` header.

I'm all in for using built-ins to get id/size information. 
@agozillon, feel free to propose any changes simplifying triSYCL development. I think we can re-use https://github.com/triSYCL/sycl/blob/sycl/unified/next/sycl/include/CL/__spir/spir_vars.hpp and have two different implementations of `__spir_ocl_get_*` built-ins:
```c++
#if __COMPILE_TO_SPIRV__
typedef size_t size_t_vec __attribute__((ext_vector_type(3)));
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalSize;
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalInvocationId;
extern "C" const __constant size_t_vec __spirv_BuiltInWorkgroupSize;
extern "C" const __constant size_t_vec __spirv_BuiltInLocalInvocationId;
extern "C" const __constant size_t_vec __spirv_BuiltInWorkgroupId;
extern "C" const __constant size_t_vec __spirv_BuiltInGlobalOffset;

// __spir_ocl_get_* implementations can be generated by macro function
constexpr __spir_ocl_get_global_id(int dim) {
  switch (dim) {
  case 0: 
    return __spirv_BuiltInGlobalInvocationId.x;
  case 1: 
    return __spirv_BuiltInGlobalInvocationId.y;
  case 2: 
    return __spirv_BuiltInGlobalInvocationId.z;
  }
}
...
#elif __XILINX_SPECIFIC_IMPLEMENTATION__
  // some other implementation of __spir_ocl_get_*
#endif
```

kbobrovs replied:
@bader, I already made the change to device_rt_intrin.hpp, which seems to address @agozillon's concerns. Are you OK if we live with this name for now?

bader replied:
This PR became a little bit messy. Now first commit adds spirv_glue.hpp file, which is later renamed to device_rt_intrin.hpp in the third commit of the pull request. 
Now I have a few options how to merge this pull request, but none of them look good to me. I can squash all the commits into single commit, but in this case first two commits (not directly related to the feature) will be part of the hierarchical parallelism commit. Another option is to save current history with merge or rebase, but it also looks strange because first patch comment about renaming is addressed in the third commit.

I really like the separation of first two commits because it's makes review easier and keeps history clean.
@kbobrovs, could you send first two commits not related to hierarchical parallelism in a separate pull request, please?
NOTE: it means that file with `init*` should be named device_rt_intrin.hpp in the first commit.

kbobrovs replied:
@bader, OK, will do

bader replied:
Thank you!

agozillon replied:
Something along those lines would be useful I think, I don't necessarily mind what form it takes and the top level function doesn't need to be named __spir_ocl_get_*  it could be __sycl_get_* if you wish, in our case that's just my cheap/cheat way of working around conflicts with user functions with the same mangling as OpenCL SPIR built-ins like get_global_id etc.

If you guys are happy with an abstraction like that (similar to our current approach) it's good with us, but perhaps someone has a more generic approach? As for the most part it would be nice to have an API for these functions that allows anyone to add there own underlying implementation, not necessarily just SPIR/SPIRV built-ins!

I had a slightly more generic and perhaps less sane implementation in a previous commit: https://github.com/triSYCL/sycl/commit/489f98d16adb07a3b4a805865b348b0ed8a6d52b#diff-0e3fe0cefdf466dc994182d2e123ba01 that may be more useful? I reverted it as we don't really have need for other built-ins just yet, so I can't really justify forcing someone else to maintain it... and a separate file approach allowed me to avoid frequent merge conflicts.

In either case, I don't want to be a bother and hold up these commits for too long, it can perhaps be a separate commit/merge/issue discussion. For now I'm happy with either, the renaming of the header to something less SPIRV centric or the movement of the functions to reside in spirv_vars. Whatever works for you guys!

kbobrovs replied:
Introducing _sycl_get* (or __sycl_get*) C APIs (C  mangling assumed, right?) sounds like a good idea. Those would be generic SYCL "intrinsics" which must be resolved by actual  SYCL device Back-end, e.g. by linking with <be>_device.bc which will provide implementations of __sycl_get* for <be>. So far we have __spirv* serving in this role, with a nice side effect that we don't need to provide <spirv>_device.bc, because SPIRV translator knows how to lower them. But the problem with current __spirv* is that the interface is based on variables rather than functions, which is less robust.

This sounds like a separate important task which needs little more discussion, so I suggest that for the purpose of hierarchical parallelism I follow what @bader  suggested - move init functions in the spirv_vars.hpp for now removing the device_rt_intrin.hpp, then we can refactor along the lines @agozillon proposed.
?

agozillon replied:
Yes, that sounds like a separate task. I am happy with the change of removing the new header and moving the functions into spirv_vars.hpp for now, that seems fine.

bader commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -89,6 +89,19 @@ decltype(member_ptr_helper(&F::operator())) argument_helper(F);
+// types: if \c Name is undefined (is a \c std::nullptr_t) then \c Type becomes
+// the \c Name.
+template <typename Name, typename Type> struct get_kernel_name_t {
+  typedef Name name;
```
Why not `using`?

kbobrovs replied:
OK

bader commented on sycl/include/CL/sycl/handler.hpp:
```diff
@@ -89,6 +89,19 @@ decltype(member_ptr_helper(&F::operator())) argument_helper(F);
 template <typename T>
 using lambda_arg_type = decltype(argument_helper(std::declval<T>()));
+
+// Helper struct to get a kernel name type based on given \c Name and \c Type
```
Are Doxygen markings applied to regular comment blocks?
Shouldn't they have to be "special comment blocks"?
http://www.doxygen.nl/manual/docblocks.html#cppblock

kbobrovs replied:
OK

keryell commented on sycl/include/CL/sycl/item.hpp:
```diff
@@ -20,6 +20,8 @@ struct Builder;
 }
 template <int dimensions> struct id;
 template <int dimensions> class range;
+template <int dimensions> struct h_item;
```
Probably everything is a `class` instead of a `struct` now

kbobrovs replied:
ok, will make h_item a class.

keryell commented on sycl/include/CL/sycl/item.hpp:
```diff
@@ -20,6 +20,8 @@ struct Builder;
 }
 template <int dimensions> struct id;
```
Probably everything is a `class` instead of a `struct` now

kbobrovs replied:
id is still a struct

keryell commented on sycl/include/CL/sycl/item.hpp:
```diff
@@ -86,6 +88,7 @@ template <int dimensions = 1, bool with_offset = true> struct item {
   // For call constructor inside conversion operator
   friend struct item<dimensions, false>;
```
Probably everything along is a `class` instead of a `struct` now

kbobrovs replied:
item is still a struct

keryell replied:
It is always good to have some feedback from the implementors.
If you find some discrepancies in the spec, feel free to file an issue or even better a PR :-) on https://github.com/KhronosGroup/SYCL-Docs
Note that the published PDF might be out-dated compared to the source files.

keryell commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+  return err_cnt == 0;
+}
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
```
`std::min`

kbobrovs replied:
That caused some problems with compilation, which I preferred to workaround at that moment. Will see if I can make std::min work.

keryell replied:
Perhaps `std::min` was not working because it is `constexpr` only since C++14...
If so, `cl::sycl::min` might have the same limitation...

bader replied:
I'm not sure what is the problem, but maybe it was caused by the headers we included on the device and which was removed by https://github.com/intel/llvm/pull/229.
@kbobrovs, could you rebase you patch to the latest version the branch and check if you still have issues with using `std::min`, please?

kbobrovs replied:
Yes, will try.

kbobrovs replied:
cl::sycl::min worked fine - this seems to be the right choice here, thanks for the suggestion

keryell commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+int main() {
+  const int N_WG = 7;
```
All these `constexpr`

kbobrovs replied:
ok

keryell commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+      // - h_item::get_global_id
+      // The global size is not known, so kernel code needs to adapt
+      // dynamically, hence the complex loop bound and index computation.
+      std::memset(ptr, 0, range_length * sizeof(ptr[0]));
```
At least it is useful to have some tests using C-style and pointers to be sure SYCL can work with C programs... :-)

kbobrovs replied:
Right. Let's say the test aims to be closer to what usual SYCL developer would write or port :-)

keryell replied:
You are depressing me ;-)

kbobrovs commented on sycl/include/CL/sycl/group.hpp:
```diff
@@ -9,6 +9,9 @@
 #include <CL/__spirv/spirv_ops.hpp>
+#include <CL/__spirv/spirv_vars.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/spirv_glue.hpp>
```
OK

kbobrovs commented on sycl/include/CL/sycl/detail/common.hpp:
```diff
@@ -113,6 +113,96 @@ const cl_mem_fence_flags CLK_GLOBAL_MEM_FENCE  = 0x02;
+  // Terminal part of the implementation.
+  template <int DIM, template <int> class LoopBoundTy, typename FuncTy,
+            template <int> class LoopIndexTy = LoopBoundTy>
+  static ALWAYS_INLINE detail::enable_if_t<DIM == 0> iterate_impl(
```
Right, this would be more elegant.

kbobrovs replied:
@keryell, tried this. Automatic template parameter deduction does not work  well for constructors in C++11, so I'd better leave std::enable_if version, as it requires much less typing on the user side. Any comments /objections?

kbobrovs replied:
Interface remained the same, implementation will now use struct specialization for DIM==0 instead of enable_if_t to terminate recursion

kbobrovs commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -45,11 +45,12 @@ class ArgDesc {
       GlobalSize[I] = 1;
       LocalSize[I] = LocalSize[0] ? 1 : 0;
       GlobalOffset[I] = 0;
+      NumWorkGroups[I] = 0;
```
This sounds interesting. Maybe it is worth to open an issue and discuss the redesign there? @romanovvlad - please chime in.

This patch just follows the existing design here.

romanovvlad replied:
We always use 3 dimensions to get rid of template parameters, so we can easily pass it to the sycl library. We could probably create 3 different types for 3 dimensions, all of these types would be inherited from common interface, but I think it's too complex for this simple structure.

keryell replied:
I understand your point.
On the long term we can do some optimization if we realize that it means more resources on the device to store this useless information and so on.

romanovvlad replied:
These fields are stored on host side only.

keryell replied:
Ah, not that bad, then.

kbobrovs commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -217,6 +272,30 @@ class HostKernel : public HostKernelBase {
     }
   }
+
+  template <class ArgT = KernelArgType>
```
OK

kbobrovs commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -217,6 +272,30 @@ class HostKernel : public HostKernelBase {
   }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, group<Dims>>::value>::type
```
Oh, right

bader commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+  if (passed) {
+    std::cout << "Passed\n";
+    return 0;
+  } else {
```
Please, remove `else`.

kbobrovs replied:
ok

bader commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -245,6 +246,37 @@ cl_int MemCpyCommandHost::enqueueImp() {
 
+// SYCL has a parallel_for_work_group variant where the only NDRange
+// characteristics set by a user is the number of work groups. This does not map
+// to the OpenCL clEnqueueNDRangeAPI, which requires global work size to be set
```
I think it's better to refer to the SYCL specification (from 4.8.5.3):
> The member function handler::parallel_for_work_group is parameterized by the number of work-groups, such that
_**the size of each group is chosen by the runtime**_, or by the number of work-groups and number of work-items for
users who need more control.

kbobrovs replied:
I think my explanation gives a clearer explanation of the motives of this function and what it does. I will add this spec reference as additional info.

bader commented on sycl/source/detail/scheduler/commands.cpp:
```diff
@@ -245,6 +246,37 @@ cl_int MemCpyCommandHost::enqueueImp() {
+  assert(NDR.NumWorkGroups[0] != 0 && NDR.LocalSize[0] == 0);
+  range<3> WGSize = get_kernel_work_group_info_cl<
+      range<3>,
+      cl::sycl::info::kernel_work_group::compile_work_group_size>::_(Kernel,
```
To improve performance we should cache this information in SYCL runtime instead of requesting it through lower level API, but this is a separate task. Could you add TODO here, please?

kbobrovs replied:
ok.

bader commented on sycl/include/CL/sycl/id.hpp:
```diff
@@ -33,8 +33,9 @@ template <int dimensions = 1> struct id : public detail::array<dimensions> {
-  template <int N = dimensions>
-  id(typename std::enable_if<(N == 1), const item<dimensions> &>::type item)
+  template <int N = dimensions, bool with_offset = true>
+  id(typename std::enable_if<(N == 1),
```
I think `enable_if_t` can be used in this file.

kbobrovs replied:
ok

kbobrovs commented on sycl/include/CL/sycl/detail/common.hpp:
```diff
@@ -113,6 +113,96 @@ const cl_mem_fence_flags CLK_GLOBAL_MEM_FENCE  = 0x02;
+
+// Fills the lack of enable_if_t in C++11.
+template <bool Cond, typename T = void>
+using enable_if_t = typename std::enable_if<Cond, T>::type;
```
ok.

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+  const int WG_SIZE_MAIN = 3;
+  const int WG_SIZE1 = 5;
+  const int WG_SIZE2 = 2;
+  const int N_ITER = 1000;
```
ok

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+      buffer<int, 1> buf(ptr, range<1>(range_length));
+      myQueue.submit([&](handler &cgh) {
+        auto dev_ptr = buf.get_access<access::mode::read_write>(cgh);
+        // numer of buf elements per work group:
```
buf here is a variable name, so I'd rather leave it as is

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+//===----------------------------------------------------------------------===//
+
+// RUN: %clang -std=c++11 -fsycl %s -o %t.out -lOpenCL -lstdc++
+// RUN: %t.out
```
ok

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <memory>
```
ok

kbobrovs replied:
Applied - this place did not change. Am I missing something?

bader replied:
> Applied - this place did not change. Am I missing something?

Probably not. I can't remember whether it's the exact place, which caught my eye, but if you applied clang-format to the whole patch, it should fix all the formatting issues. Thanks!

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+
+        cgh.parallel_for_work_group<class hpar_simple>(
+            range<1>(N_WG), [=](group<1> g) {
+              size_t wg_off = wg_chunk * g.get_id(0);
```
"off" is a widely used acronym for "offset" so I'd leave the var name as is

bader replied:
There is even more widely used English word "off" and I spend some time trying to understand why and how we disable work-group.
What is the problem with replacing acronym with the full word?

kbobrovs replied:
Were troubles with understanding related to "off" vs "offset"? If so, I'll be happy to change, no problems.

bader replied:
Yes. Thanks!

kbobrovs replied:
ok - will update then

kbobrovs commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -72,14 +85,42 @@ class NDRDescT {
+  /// Number of workgroups, used to record the number of workgroups from the
+  /// simplest form of parallel_for_work_group. If set, all other fields must be
+  /// zero
+  sycl::range<3> NumWorkGroups;
```
as @romanovvlad wrote:
> We always use 3 dimensions to get rid of template parameters

kbobrovs commented:
As suggested by @bader, I split this PR into two:
https://github.com/intel/llvm/pull/239
https://github.com/kbobrovs/llvm/pull/1

The last one depends on the previous one, which is not yet in intel/llvm, so I could not create it in intel/llvm. I'm was not sure how to best handle this, so I added all reviewers as collaborators into my repo, so that you guys can review there. I hope I addressed all outstanding comments from this PR.
@keryell, @agozillon, @romanovvlad - please take another look.

kbobrovs commented:
Please review the last two commits (not sure why other commits got into this PR):
- [SYCL] Move getNextPowerOfTwo body into the header to avoid link errors.
   -- this one fixes some existing problem which prevented me from testing
- [SYCL] Implement hierarchical parallelism API.
   -- this one is moved from my private branch PR (https://github.com/kbobrovs/llvm/pull/1) back here per @bader 's request. All comments are addressed.

kbobrovs commented:
> [SYCL] Move getNextPowerOfTwo body into the header to avoid link errors.
> -- this one fixes some existing problem which prevented me from testing

This one was fixed by others, so please review only 1 commit - [SYCL] Implement hierarchical parallelism API.

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+      buffer<int, 1> buf(ptr, range<1>(range_length));
+      myQueue.submit([&](handler &cgh) {
+        auto dev_ptr = buf.get_access<access::mode::read_write>(cgh);
+        // numer of 'buf' elements per work group:
```
ok

kbobrovs commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,183 @@
+  constexpr int N_WG = 7;
+  constexpr int WG_SIZE_MAIN = 3;
+  constexpr int WG_SIZE1 = 5;
+  constexpr int WG_SIZE2 = 2;
```
They are both logical. But I see your point  - will invent a better name

bader replied:
I think `WG_SIZE_MAIN` is not logical.

kbobrovs replied:
Right. WG_SIZE1 and WG_SIZE2 are. Updated

kbobrovs commented on sycl/include/CL/sycl/detail/cg.hpp:
```diff
@@ -102,7 +147,25 @@ class HostKernel : public HostKernelBase {
-  void call(const NDRDescT &NDRDesc) override { runOnHost(NDRDesc); }
+  void call(const NDRDescT &NDRDesc) override {
+    // adjust ND range for serial host:
+    NDRDescT R1;
```
ok

kbobrovs commented on sycl/include/CL/sycl/group.hpp:
```diff
@@ -66,12 +83,113 @@ template <int dimensions = 1> class group {
+    range<dimensions> global_size;
+    range<dimensions> local_size;
+    id<dimensions> global_id;
+    id<dimensions> local_id;
```
ok

kbobrovs replied:
@bader, looks like you listed all unresolved comments from the previous review here, so should be fixed now.

bader commented on sycl/test/hier_par/hier_par_basic.cpp:
```diff
@@ -0,0 +1,185 @@
+      buffer<int, 1> buf(ptr, range<1>(range_length));
+      myQueue.submit([&](handler &cgh) {
+        auto dev_ptr = buf.get_access<access::mode::read_write>(cgh);
+        // numer of 'buf' elements per work group:
```
```suggestion
        // nubmer of 'buf' elements per work group:
```

kbobrovs replied:
Oops. Thought I'd fixed it. Thanks

kbobrovs commented:
Had to revert template <... template <int> typename ...> back to template <... template <int> class ...>, because the former leads to warning that it is C++17 feature and one of test failures.

bader commented:
@kbobrovs, could you resolve the conflict with recently committed patches, please?

kbobrovs commented:
> @kbobrovs, could you resolve the conflict with recently committed patches, please?

ok, done
